modules/renter/streamcache.go

```diff
@@ -0,0 +1,181 @@
+	cd := &chunkData{
+		id:         cacheID,
+		data:       data,
+		lastAccess: time.Now(),
+	}
+	sc.streamMap[cacheID] = cd
+	heap.Push(&sc.streamHeap, cd)
+	sc.streamHeap.update(cd, cd.id, cd.data, cd.lastAccess)
+}
+
+// init initializes the streamCache
+func (sc *streamCache) Init() {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	if sc.cacheSize > 0 {
+		build.Critical("streamCache already initialized")
+		return
+	}
+
+	sc.streamMap = make(map[string]*chunkData)
+	sc.streamHeap = make(streamHeap, 0, defaultStreamCacheSize)
+	sc.cacheSize = defaultStreamCacheSize
+	heap.Init(&sc.streamHeap)
+}
+
+// pruneCache prunes the cache until it is the length of size
+func (sc *streamCache) pruneCache(size uint64) {
+	for len(sc.streamMap) > int(size) {
+		// Remove from Heap
+		cd := heap.Pop(&sc.streamHeap).(*chunkData)
+
+		// Remove from Map
+		if _, ok := sc.streamMap[cd.id]; !ok {
+			build.Critical("Cache Data chunk not found in streamMap.")
+		}
+		delete(sc.streamMap, cd.id)
+	}
+}
+
+// Retrieve tries to retrieve the chunk from the renter's cache. If
+// successful it will write the data to the destination and stop the download
+// if it was the last missing chunk. The function returns true if the chunk was
+// in the cache.
+// Using the entire unfisihedDownloadChunk as the argument as there are seven different fields
+// used from unfinishedDownloadChunk and it allows using udc.fail()
+//
+// TODO: in the future we might need cache invalidation. At the
+// moment this doesn't worry us since our files are static.
+func (sc *streamCache) Retrieve(udc *unfinishedDownloadChunk) bool {
+	udc.mu.Lock()
+	defer udc.mu.Unlock()
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	cd, cached := sc.streamMap[udc.staticCacheID]
+	if !cached {
+		return false
+	}
+
+	// chunk exists, updating lastAccess and reinserting into map, updating heap
+	cd.lastAccess = time.Now()
+	sc.streamMap[udc.staticCacheID] = cd
+	sc.streamHeap.update(cd, cd.id, cd.data, cd.lastAccess)
+
+	start := udc.staticFetchOffset
+	end := start + udc.staticFetchLength
+	_, err := udc.destination.WriteAt(cd.data[start:end], udc.staticWriteOffset)
+	if err != nil {
+		udc.fail(errors.AddContext(err, "failed to write cached chunk to destination"))
+		return true
+	}
+
+	// Check if the download is complete now.
+	udc.download.mu.Lock()
+	defer udc.download.mu.Unlock()
+
+	udc.download.chunksRemaining--
+	if udc.download.chunksRemaining == 0 {
+		udc.download.endTime = time.Now()
+		close(udc.download.completeChan)
+		udc.download.destination.Close()
+		udc.download.destination = nil
+	}
+	return true
+}
+
+// SetStreamingCacheSize confirms that the cache size is being set
+// to a value greater than zero.  Otherwise it will remain the default
+// value set during the initialization of the streamCache.
+// It will also prune the cache to ensure the cache is always
+// less than or equal to whatever the cacheSize is set to
+func (sc *streamCache) SetStreamingCacheSize(cacheSize uint64) {
```

Comment:
this behavior is a little unusual. I think I'd prefer to punt the `cacheSize > 0` check up to the caller of `SetStreamingCacheSize`.